pkg/steps/censoringclient/client.go

```diff
@@ -0,0 +1,138 @@
+package censoringclient
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+
+	"github.com/openshift/ci-tools/pkg/secrets"
+	"github.com/openshift/ci-tools/pkg/util/watchingclient"
+)
+
+func Wrap(upstream watchingclient.Client, censor *secrets.DynamicCensor) watchingclient.Client {
+	return &client{
+		upstream: upstream,
+		censor:   censor,
+	}
+}
+
+// client updates the censor when it interacts with secrets on the cluster
+type client struct {
+	upstream watchingclient.Client
+	censor   *secrets.DynamicCensor
+}
+
+func (c *client) Get(ctx context.Context, key ctrlruntimeclient.ObjectKey, obj ctrlruntimeclient.Object) error {
+	if err := c.upstream.Get(ctx, key, obj); err != nil {
+		return err
+	}
+	c.record(obj)
+	return nil
+}
+
+func (c *client) List(ctx context.Context, list ctrlruntimeclient.ObjectList, opts ...ctrlruntimeclient.ListOption) error {
+	if err := c.upstream.List(ctx, list, opts...); err != nil {
+		return err
+	}
+	c.recordList(list)
+	return nil
+}
+
+func (c *client) Create(ctx context.Context, obj ctrlruntimeclient.Object, opts ...ctrlruntimeclient.CreateOption) error {
+	if err := c.upstream.Create(ctx, obj, opts...); err != nil {
+		return err
+	}
+	c.record(obj)
+	return nil
+}
+
+func (c *client) Delete(ctx context.Context, obj ctrlruntimeclient.Object, opts ...ctrlruntimeclient.DeleteOption) error {
+	return c.upstream.Delete(ctx, obj, opts...)
+}
+
+func (c *client) Update(ctx context.Context, obj ctrlruntimeclient.Object, opts ...ctrlruntimeclient.UpdateOption) error {
+	if err := c.upstream.Update(ctx, obj, opts...); err != nil {
+		return err
+	}
+	c.record(obj)
+	return nil
+}
+
+func (c *client) Patch(ctx context.Context, obj ctrlruntimeclient.Object, patch ctrlruntimeclient.Patch, opts ...ctrlruntimeclient.PatchOption) error {
+	if err := c.upstream.Patch(ctx, obj, patch, opts...); err != nil {
+		return err
+	}
+	c.record(obj)
+	return nil
+}
+
+func (c *client) DeleteAllOf(ctx context.Context, obj ctrlruntimeclient.Object, opts ...ctrlruntimeclient.DeleteAllOfOption) error {
+	return c.upstream.DeleteAllOf(ctx, obj, opts...)
+}
+
+func (c *client) Watch(ctx context.Context, obj ctrlruntimeclient.ObjectList, opts ...ctrlruntimeclient.ListOption) (watch.Interface, error) {
+	return c.upstream.Watch(ctx, obj, opts...)
+}
+
+func (c *client) recordList(obj ctrlruntimeclient.ObjectList) {
+	gvk, err := apiutil.GVKForObject(obj, c.Scheme())
```

Comment:
This assumes the type you have is registered in the scheme.  I didn't read to see if that is always true.

If it is possible that the type is not registered in the scheme, you may need to inspect the GVK using GetObjectKind to handle unstructured.